### PR TITLE
🚑 Fix & Chore: Fatal entry point error, test, and type declarations

### PR DESCRIPTION
**Pull request by HeavenlyEntity**
# Changes

### ➕ Added
- Added Jest config file for proper test running and local testing.
- Added babel config to all for ES6 imports on jest tests and enable typescript support

### 🐛 Fixes
- Set up tests and new node version 18.12.0 support #583 
- Entry error induced by new node version 
- Linter locating, and fix application scripts 
- Fixed type declarations and organized them for scalability. 

### 📦 Updates
- Updated `jest` & `ts-jest` to **v29.5**

Currently trying to use this for a project 😆 -- Then had this:
```shell
Failed to resolve entry for package "@octokit/plugin-throttling". The package may have incorrect main/module/exports specified in its package.json.
```

## Pull request type
Please add the corresponding label for change this PR introduces:

Bugfix: `Type: Bug`
Feature/model/API additions: `Type: Feature`
Updates to docs or samples: `Type: Documentation`
Dependencies/code cleanup: `Type: Maintenance`

### CODE REVIEW

**Comment by wolfy1339** [comment]
Why do we need Babel? And why is the Jest config in `package.json` not sufficient?

Why do we need a default export in `src/index.ts`? We don't use default exports in the Octokit projects

There isn't many details as to the why of this PR, only the what changed

**Comment by HeavenlyEntity** [comment]
Hey @wolfy1339 

After the latest version of node I could not get the tests to run nor npm to install packages properly. `pnpm` was introduced to allow backward compatibility moving forward. When trying to run tests previously the `lint` script was not finding files. Also babel is used to allow the ES6 style imports to run the tests without having to import via `require` as function mocks were throwing errors requiring "module.export" to work with the latest syntax. Babel fixes that. The `jest.config` helps to map the files/directories to be tested and excluded. This helps extremely for scalability, readability, and maintainability.

File: package.json [hunk 1]
```diff
@@ -27,35 +27,31 @@
...
+    "jest": "^29.5.0",
     "npm-run-all": "^4.1.5",
     "prettier": "2.8.7",
     "semantic-release-plugin-update-version-in-files": "^1.0.0",
-    "ts-jest": "^29.0.0",
+    "ts-jest": "^29.1.0",
+    "ts-node": "^10.9.1",
     "typescript": "^5.0.0"
   },
-  "jest": {
```
Comment (HeavenlyEntity) [inline package.json:48]: When running the `jest` command it was throwing an error and running the tests in the `src` folder. Completely ignoring the `tests` folder after updating the jest version.  `jest.config.ts` replaced this.

File: package.json [hunk 2]
```diff
@@ -7,14 +7,14 @@
   "description": "Octokit plugin for GitHub's recommended request throttling",
   "scripts": {
     "build": "pika-pack build",
-    "lint": "prettier --check '{src,scripts,test}/**/*' '!*/generated/**' README.md package.json",
-    "lint:fix": "prettier --write '{src,scripts,test}/**/*' '!*/generated/**' README.md package.json",
+    "lint": "prettier --check \"src/**/*.ts\" \"!*/generated/**\" README.md package.json",
+    "lint:fix": "prettier --write  \"src/**/*.ts\" \"!*/generated/**\" README.md package.json",
     "pretest": "npm run -s lint",
-    "test": "jest --coverage",
```
Comment (HeavenlyEntity) [inline package.json:13]: Coverage is already included in config options.

**Comment by HeavenlyEntity** [comment]
@wolfy1339 so found `npm` to be pushing bad releases. Which brings problems to users using `nvm`. Each and every time I use `npm` - `v8.19.2`. I get `ERR! Unexpected token '.'`. So note `v9.5.1` this issue don't happen with Node version `18.16.0`. 

Removing need for `pnpm`

**Comment by wolfy1339** [comment]
> After the latest version of node I could not get the tests to run nor npm to install packages properly. `pnpm` was introduced to allow backward compatibility moving forward.

That is odd. What exact versions are you using?
I don't have any issues with installing any dependencies, and the tests run just fine.

```
$ node --version
v18.15.0
$ npm --version
9.5.0
```
I even tried NodeJS 20, with the same result

```
$ node --version
v20.0.0
$ npm --version
9.6.4
```

> When trying to run tests previously the `lint` script was not finding files.

What environment are you on? Can you check if changing the single quotes to double quotes fixes the issue, like in this commit https://github.com/octokit/rest.js/commit/6dde9ed7b3b1055f6a1bc7458776aae89451205a and leave the rest of the script untouched

> Also babel is used to allow the ES6 style imports to run the tests without having to import via `require` as function mocks were throwing errors requiring "module.export" to work with the latest syntax. Babel fixes that.

I cannot reproduce this.
Where are you seeing `require`? The `tsconfig.json` has the `module` option set to `esnext`, which means that it will output as an ESModule, and won't use `require`

> The `jest.config` helps to map the files/directories to be tested and excluded. This helps extremely for scalability, readability, and maintainability.

The `ts-jest` preset is the only thing we need, and works just fine. We don't like to have configuration files if we can avoid to.

**Pull request closed without merging**